spark/src/main/scala/org/apache/spark/sql/delta/TypeWideningMetadata.scala

```diff
@@ -171,23 +175,48 @@ private[delta] object TypeWideningMetadata extends DeltaLogging {
     schemaWithMetadata
   }
 
+  /**
+   * Recursively compare `from` and `to` to collect all primitive type widening changes, including
+   * in nested structs, maps and arrays.
+   */
+  def collectTypeChanges(from: StructType, to: StructType): Seq[TypeChange] = {
```

Comment:
Nit: It's a bit confusing to have two methods called `collectTypeChanges`. Why couldn't we add a `case (from: StructType, to: StructType)` to the other one?

Reply:
I renamed the internal collectTypeChanges to collectTypeChangesInStructField to better distinguish them.

We do need collectTypeChangesInStructField that doesn't recurse into structs to be able to map all type changes to their nearest struct field ancestor, which is where we attach the metadata recording the history of type changes within that struct field (to go a bit deeper: this is due to the fact that we can't attach metadata to map key/values and array elements, so we need collect type changes inside these while ignoring structs)